### PR TITLE
fix: ensure node leases aren't leaked

### PR DESCRIPTION
**Pull request by jmdeal**
Fixes #N/A 

**Description**

This PR makes two changes to the NodeClaim lifecycle controller to ensure node leases aren't leaked:
- The NodeClaim must be registered before being initialized. This ensures NodeClaims aren't candidates for voluntary disruption mechanisms if the Node hasn't had the termination finalizer added by the registration sub-reconciler.
- When terminating a NodeClaim, only delete the associated Nodes if the NodeClaim is registered. Similar to the first change, this ensures we don't delete Nodes which don't have Karpenter's termination finalizer. These Nodes should be removed out-of-band by CCM once the underlying instance has terminated.

**How was this change tested?**
`make test`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by njtran — COMMENTED** [review]
/lgtm
/approve

File: pkg/controllers/nodeclaim/lifecycle/initialization.go [hunk 1]
```diff
@@ -45,6 +45,9 @@ type Initialization struct {
 // c) all extended resources have been registered
 // This method handles both nil nodepools and nodes without extended resources gracefully.
 func (i *Initialization) Reconcile(ctx context.Context, nodeClaim *v1.NodeClaim) (reconcile.Result, error) {
+	if !nodeClaim.StatusConditions().Get(v1.ConditionTypeRegistered).IsTrue() {
```
[the remark this answers is not in the document]
Comment (jmdeal) [reply pkg/controllers/nodeclaim/lifecycle/initialization.go]: Agreed, I've made the update.

**Review by jonathan-innis — APPROVED** [review]
/lgtm
/approve

**Pull request merged**